storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -105,134 +93,99 @@ impl DBPruner for StateStorePruner {
 }
 
 impl StateStorePruner {
-    pub fn new(db: Arc<DB>, max_index_purged_version: Version, index_purged_at: Instant) -> Self {
+    pub fn new(db: Arc<DB>) -> Self {
         let pruner = StateStorePruner {
             db,
-            max_index_purged_version: AtomicVersion::new(max_index_purged_version),
-            index_purged_at: Mutex::new(index_purged_at),
             target_version: AtomicVersion::new(0),
             min_readable_version: AtomicVersion::new(0),
         };
         pruner.initialize();
         pruner
     }
 
-    /// Purge the stale node index so that after restart not too much already pruned stuff is dealt
-    /// with again (although no harm is done deleting those then non-existent things.)
-    ///
-    /// We issue (range) deletes on the index only periodically instead of after every pruning batch
-    /// to avoid sending too many deletions to the DB, which takes disk space and slows it down.
-    fn maybe_purge_index(&self) -> Result<()> {
-        const MIN_INTERVAL: Duration = Duration::from_secs(10);
-        const MIN_VERSIONS: u64 = 60000;
-
-        let min_readable_version = self.min_readable_version.load(Ordering::Relaxed);
-        let max_index_purged_version = self.max_index_purged_version.load(Ordering::Relaxed);
-
-        // A deletion is issued at most once in one minute and when the pruner has progressed by at
-        // least 60000 versions (assuming the pruner deletes as slow as 1000 versions per second,
-        // this imposes at most one minute of work in vain after restarting.)
-        let now = Instant::now();
-        if now - *self.index_purged_at.lock() > MIN_INTERVAL
-            && min_readable_version - max_index_purged_version > MIN_VERSIONS
-        {
-            self.db.range_delete::<StaleNodeIndexSchema, Version>(
-                &(max_index_purged_version + 1), // begin is inclusive
-                &(min_readable_version + 1),     // end is exclusive
-            )?;
-            // The index items at min_readable_version has been consumed already because they
-            // indicate nodes that became stale at that version, keeping that version readable,
-            // hence is purged above.
-            self.max_index_purged_version
-                .store(min_readable_version, Ordering::Relaxed);
-            *self.index_purged_at.lock() = now;
+    pub fn prune_state_store(
+        &self,
+        min_readable_version: Version,
+        target_version: Version,
+        batch_size: usize,
+    ) -> anyhow::Result<Version> {
+        let indices_iter = StaleNodeIndicesByNumberOfItemsIterator::new(
+            &self.db,
+            min_readable_version,
+            batch_size,
+            target_version,
+        )?;
+        let indices = indices_iter.collect::<anyhow::Result<Vec<_>>>()?;
+
+        if indices.is_empty() {
+            Ok(min_readable_version)
```

Comment:
should be `target_version`, learnt it the hard way.

https://github.com/aptos-labs/aptos-core/pull/2007

Reply:
Done.